Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m

```diff
@@ -0,0 +1,298 @@
+//
+//  TDURLConnectionChangeTracker.m
+//  
+//
+//  Created by Adam Cox on 1/5/15.
+//  Copyright (c) 2015 IBM.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+//  except in compliance with the License. You may obtain a copy of the License at
+//    http://www.apache.org/licenses/LICENSE-2.0
+//  Unless required by applicable law or agreed to in writing, software distributed under the
+//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+//  either express or implied. See the License for the specific language governing permissions
+//  and limitations under the License.
+//
+// <http://wiki.apache.org/couchdb/HTTP_database_API#Changes>
+//
+
+#import "TDURLConnectionChangeTracker.h"
+#import "TDRemoteRequest.h"
+#import "TDAuthorizer.h"
+#import "TDStatus.h"
+#import "TDBase64.h"
+#import "MYBlockUtils.h"
+#import "MYURLUtils.h"
+#import <string.h>
+#import "TDJSON.h"
+#import "CDTLogging.h"
+#import "TDMisc.h"
+
+#define kMaxRetries 6
+#define kInitialRetryDelay 0.2
+
+@interface TDURLConnectionChangeTracker()
+@property (strong, nonatomic) NSMutableData* inputBuffer;
+@property (strong, nonatomic) NSURLConnection *connection;
+@property (strong, nonatomic) NSMutableURLRequest *request;
+@property (strong, nonatomic) NSDate* startTime;
+@property (nonatomic, readwrite) NSUInteger totalRetries;
+@end
+
+@implementation TDURLConnectionChangeTracker
+
+- (BOOL)start
+{
+    if (self.connection) return NO;
+    
+    CDTLogInfo(CDTREPLICATION_LOG_CONTEXT, @"%@: Starting...", [self class]);
+    [super start];
+    
+    NSURL* url = self.changesFeedURL;
+    self.request = [[NSMutableURLRequest alloc] initWithURL:url];
+    self.request.cachePolicy = NSURLRequestReloadIgnoringLocalCacheData;
+    self.request.HTTPMethod = @"GET";
+    
+    // Add headers from my .requestHeaders property:
+    for(NSString *key in self.requestHeaders) {
+        [self.request setValue:self.requestHeaders[key] forHTTPHeaderField:key];
+    }
+    
+    // Add cookie headers from the NSHTTPCookieStorage:
+    NSArray* cookies = [[NSHTTPCookieStorage sharedHTTPCookieStorage] cookiesForURL:url];
+    NSDictionary* cookieHeaders = [NSHTTPCookie requestHeaderFieldsWithCookies:cookies];
+    for (NSString* headerName in cookieHeaders) {
+        [self.request setValue:cookieHeaders[headerName] forHTTPHeaderField:headerName];
+    }
+
+    if (self.authorizer) {
+        NSString* authHeader = [self.authorizer authorizeURLRequest:self.request forRealm:nil];
+        if (authHeader) {
+            [self.request setValue: authHeader forHTTPHeaderField:@"Authorization"];
+        }
+    }
```

Comment:
For both the cookie headers and this authorization header addition, issue a warning log line if you are overwriting something already in the custom headers passed to the request. I know that at least Authorization shouldn't happen, but we might change the pre-conditions for this class in future.